nucleus/annotation.py

```diff
@@ -364,6 +370,144 @@ def to_payload(self) -> dict:
         return payload
 
 
+@dataclass
+class Keypoint:
+    """A 2D point that has an additional visibility flag.
+
+    Keypoints are intended to be part of a larger collection, and connected
+    via a pre-defined skeleton. A keypoint in this skeleton may be visible
+    or not-visible, and may be unlabeled and not visible. Because of this,
+    the x, y coordinates may be optional, assuming that the keypoint is not
+    visible, and would not be shown as part of the combined label.
+
+    Parameters:
+        x (Optional[float]): The x coordinate of the point.
+        y (Optional[float]): The y coordinate of the point.
+        visible (bool): The visibility of the point.
+    """
+
+    x: Optional[float] = None
+    y: Optional[float] = None
+    visible: bool = True
+
+    def __post_init__(self):
+        if self.visible and (self.x is None or self.y is None):
+            raise ValueError(
+                "Visible keypoints must have non-None x and y coordinates"
+            )
+
+    @classmethod
+    def from_json(cls, payload: Dict[str, Union[float, bool]]):
+        return cls(
+            payload.get(X_KEY, None),
+            payload.get(Y_KEY, None),
+            bool(payload[VISIBLE_KEY]),
+        )
+
+    def to_payload(self) -> dict:
+        return {
+            X_KEY: self.x,
+            Y_KEY: self.y,
+            VISIBLE_KEY: self.visible,
+        }
+
+
+@dataclass
+class KeypointsAnnotation(Annotation):
+    """A keypoints annotation containing a list of keypoints and the structure
+    of those keypoints: the naming of each point and the skeleton that connects
+    those keypoints.
+
+    ::
+
+        from nucleus import KeypointsAnnotation
+
+        keypoints = KeypointsAnnotation(
+            label="face",
+            keypoints=[Keypoint(100, 100), Keypoint(120, 120), Keypoint(visible=False), Keypoint(0, 0)],
+            names=["point1", "point2", "point3", "point4"],
+            skeleton=[[0, 1], [1, 2], [1, 3], [2, 4]],
```

Comment:
not super familiar with typical OOP paradigms for keypoints, but filed a long term follow up to support string based skeletons like `[["point1", "point2"], ...]`